hydroflow_lang/src/graph/ops/buffer.rs

```diff
@@ -0,0 +1,124 @@
+    hard_range_inn: RANGE_1,
+    soft_range_inn: RANGE_1,
+    hard_range_out: RANGE_1,
+    soft_range_out: RANGE_1,
+    num_args: 1,
+    is_external_input: false,
+    ports_inn: None,
+    ports_out: None,
+    properties: FlowProperties {
+        deterministic: FlowPropertyVal::Preserve,
+        monotonic: FlowPropertyVal::No,
+        inconsistency_tainted: false,
+    },
+    input_delaytype_fn: |_| None,
+    write_fn: |wc @ &WriteContextArgs {
+                   ident,
+                   op_span,
+                   root,
+                   inputs,
+                   outputs,
+                   is_pull,
+                   op_inst: OperatorInstance { arguments, .. },
+                   ..
+               },
+               _| {
+        let internal_buffer = wc.make_ident("internal_buffer");
+
+        let receiver = &arguments[0];
+        let stream_ident = wc.make_ident("stream");
+
+        let write_prologue = quote_spanned! {op_span=>
+            let mut #stream_ident = Box::pin(#receiver);
+            let #internal_buffer = df.add_state(::std::cell::RefCell::new(::std::vec::Vec::new()));
+        };
+
+        let write_iterator = if is_pull {
+            let input = &inputs[0];
+
+            quote_spanned! {op_span=>
+
+                {
+                    let mut vec = context.state_ref(#internal_buffer).borrow_mut();
+                    vec.extend(#input);
+                }
+
+                let #ident = match #root::futures::stream::Stream::poll_next(#stream_ident.as_mut(), &mut std::task::Context::from_waker(&context.waker())) {
+                    std::task::Poll::Ready(_) => {
+                        let mut vec = context.state_ref(#internal_buffer).borrow_mut();
+                        std::mem::take(&mut *vec)
+                    },
+                    std::task::Poll::Pending => {
+                        std::vec::Vec::new()
+                    },
+                }.into_iter();
+            }
+        } else {
+            let output = &outputs[0];
+
+            quote_spanned! {op_span=>
+
+                let #ident = hydroflow::pusherator::for_each::ForEach::new(|x| {
+                    let mut vec = context.state_ref(#internal_buffer).borrow_mut();
+
+                    vec.push(x);
+                });
+
+                {
+                    let mut out = #output;
+                    for x in match #root::futures::stream::Stream::poll_next(#stream_ident.as_mut(), &mut std::task::Context::from_waker(&context.waker())) {
+                        std::task::Poll::Ready(_) => {
+                            let mut vec = context.state_ref(#internal_buffer).borrow_mut();
+                            std::mem::take(&mut *vec)
+                        },
+                        std::task::Poll::Pending => {
+                            std::vec::Vec::new()
+                        },
+                    }.into_iter() {
+                        use ::pusherator::Pusherator;
```

Comment:
Is there a more explicit way to do this? It doesn't feel right to have a use statement in the generated code.

maybe something like:
<type as ::pusherator::Pusherator>::give(&mut out, x);

except we don't really know the type here only the identifier?

Reply:
oh you can just use _ as the type.

Reply:
Hmm I think you can just use `Pusherator` directly?

https://github.com/hydro-project/hydroflow/commit/688026b29490936906eb77314466eb85f95dbab3#diff-e9c19ee632ce131b963fe3a702ebb593fd1855df40cbd057a3da3ec661ca224eR115